IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,478 @@
+
+#pragma mark - Initializer
+
+-(instancetype)init
+{
+    self = [super init];
+    if (self)
+    {
+        _lastRequestTelemetry = [MSIDLastRequestTelemetry sharedInstance];
+    }
+    return self;
+}
+
+- (instancetype)initWithContext:(id<MSIDRequestContext>)context
+{
+    self = [self init];
+    if (self)
+    {
+        _context = context;
+    }
+    return self;
+}
+
+#pragma mark - Public API
+
+/**
+  Base on the thumbprint value of the request, throttling service will query database to see if any existing record and return throttling decision to calling module.
+ - NOT throttle case: return result
+ - Shoud throttle case: return cached result + update server telemetry + update cache record
+ */
+- (void)shouldThrottleRequest:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                  resultBlock:(nonnull MSIDThrottleResultBlock)resultBlock
+{
+    if ([MSIDThrottlingService validateInput:request])
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_READING_ERROR, nil);
+    }
+    
+    // If if any check return TRUE, the resultBlock will be executed inside the check, if not we return result block with shouldBeThrottle = NO
+    if ([self is429ThrottleType:request resultBlock:resultBlock] || [self isUIRequiredThrottleType:request resultBlock:resultBlock])
+    {
+        return;
+    }
+    
+    resultBlock(NO, nil);
+    return;
+}
+
+/**
+ Whenever we receives a response from server, we want to check if any throttling error to update database. This is an public API of throttling service for that task.
+ */
+- (void)updateThrottlingService:(NSError *)error tokenRequest:(id<MSIDThumbprintCalculatable>)tokenRequest
+{
+    NSError *throttlingError = nil;
+    [self updateThrottlingDatabaseWithRequest:tokenRequest
+                                errorResponse:error
+                                  returnError:&throttlingError];
+    if (throttlingError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, @"Throttling error when updating db %@, %ld", throttlingError.domain, (long)throttlingError.code);
+    }
+}
+
+#pragma mark - Internal API
+
+/**
+ return cached error if a request is 429 in db.
+ If it's a hit, we also update telemetry.
+ */
+- (BOOL)is429ThrottleType:(id<MSIDThumbprintCalculatable> _Nonnull)request
+              resultBlock:(nonnull MSIDThrottleResultBlock)resultBlock
+{
+    NSError *error = nil;
+    // Check 429 throttling case
+    NSString *strictThumbprint = [request strictRequestThumbprint];
+    MSIDThrottlingCacheRecord *cacheRecord = [self.cacheService objectForKey:strictThumbprint                                                  error:&error];
+    if (!cacheRecord)
+    {
+        // we just log error (if any) and keep moving to the next UIRequired check
+        if (error)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_READING_ERROR, error);
+        }
+        return NO;
```

Comment:
same here: no call on result block

Reply:
ah got that, if the result is a hit, we will return the block earlier. If not, we continue to check other throttling types. Eventually, if there is no hit, we will return resultBlock(NO, nil);